### PR TITLE
Modified EditForm to return _fixedEditContext via the EditContext parameter

### PR DESCRIPTION
**Pull request by chrissainty**
- Modified the `EditContext` parameter to return `_fixedEditContext`
- Introduced a new field, `_providedEditContext` which stores any `EditContext` provided by the developer. This means we don't break existing functionality.
- Updates to `OnParametersSet` to use the new `_providedEditContext` field instead of the `EditContext` property
- Unit tests to cover new `EditForm` functionality

@SteveSandersonMS tagged for review 

Addresses #12238
Addresses https://github.com/dotnet/aspnetcore/issues/23354

### CODE REVIEW

File: src/Components/Web/src/Forms/EditForm.cs
Comment (javiercn) [inline src/Components/Web/src/Forms/EditForm.cs:20]: I'm not super familiar with this area, do we need the extra field? Or can we make _providedContext hold the right value whether it received a Model or a Context

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (javiercn) [inline src/Components/Web/test/Forms/EditFormTest.cs:14]: Nice!

File: src/Components/Web/src/Forms/EditForm.cs
Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs:20]: The `_providedEditContext` field is there to track if the `EditContext` was provided directly via the `EditContext` parameter. This mean we don't break the current functionality in the `OnParametersSet` method which checks if the developer has provided a value for both `EditContext` and `Model`.

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (chrissainty) [reply src/Components/Web/test/Forms/EditFormTest.cs:14]: Thanks! I hope the tests are ok, its my first go at writing them.

**Review by javiercn — APPROVED** [review]

File: src/Components/Web/src/Forms/EditForm.cs
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/EditForm.cs:45]: I think this risks being a little weird, because I know we don't recommend people write to parameter properties directly, but if someone does this will be kind of bizarre. If you write a value to this property then try to read it back, you'll get a totally different value :)

Would it be possible instead to do something like:

```cs
bool _hasSetEditContextExplicitly;

[Parameter]
public EditContext? EditContext
{
    get => _fixedEditContext;
    set
    {
        _fixedEditContext = value;
        _hasSetEditContextExplicitly = true;
    }
}
```

... and then inside `OnParametersSet`:

```cs
if (_hasSetEditContextExplicitly && Model != null)
{
    throw "Don't set both"
}
else if (EditContext == null && Model == null)
{
   throw "Set at least one"
}
```

Sorry for not noticing this earlier on the original proposal!

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (SteveSandersonMS) [inline src/Components/Web/test/Forms/EditFormTest.cs:68]:
```suggestion
            Assert.Same(editContext, returnedEditContext);
```

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (SteveSandersonMS) [inline src/Components/Web/test/Forms/EditFormTest.cs:50]:
```suggestion
            Assert.NotNull(returnedEditContext);
            Assert.Same(model, returnedEditContext.Model);
```

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (SteveSandersonMS) [reply src/Components/Web/test/Forms/EditFormTest.cs:14]: These tests are looking great. Good stuff!

**Comment by javiercn** [comment]
This is a patch with the changes in the ref assembly, it's not an API change, but the tool is complaining, so we just need to apply that to keep the build happy
[Update-ref-assemblies.txt](https://github.com/dotnet/aspnetcore/files/4932259/Update-ref-assemblies.txt)

File: src/Components/Web/src/Forms/EditForm.cs
Comment (javiercn) [reply src/Components/Web/src/Forms/EditForm.cs:45]: This is why I can't be left alone to review things after lunch 😄

Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs:45]: THis makes sense, I'll make the change

**Comment by chrissainty** [comment]
@javiercn Is that patch something I need to do? If so what do I need to do? :)

File: src/Components/Web/src/Forms/EditForm.cs
Comment (chrissainty) [inline src/Components/Web/src/Forms/EditForm.cs:46]: I'm not sure if this was the right thing to do now I'm looking at it here. This would mean that once an EditContext has been set it can't be set back to null. Is that a scenario that should be catered for?

**Comment by javiercn** [comment]
@chrissainty you can apply the patch manually (just download it and apply the changes) if you know how, or if you do it the "manual way" inspect the patch, open the file and replace the line manually.

This is part of how we generate reference assemblies for compilation and sometimes it causes this issues. We are working on making this happen automagically I believe

File: src/Components/Web/src/Forms/EditForm.cs
Comment (javiercn) [reply src/Components/Web/src/Forms/EditForm.cs:46]: It's never going to be, is it? even if you set a model, wouldn't that be the case?

Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs:46]: I can't think of a reasonable scenario where you would pass in an EditContext and then later decide to set it to null. But just wanted to check

Comment (javiercn) [reply src/Components/Web/src/Forms/EditForm.cs:46]: I have the same feeling, but @SteveSandersonMS is the expert on this area, so I would defer the final say to him

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/EditForm.cs:46]: If you want a null check, could you throw `ArgumentNullException`? I think we shouldn't ignore the null value. Either accept it, or throw :)

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/EditForm.cs:46]: (Of these options, I'd prefer throwing, since there's no real value in letting people set it back to null)

Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs:46]: I added this check so that `_hasSetEditContextExplicity` doesn't get set to true when the value is null. Perhaps I should adjust the code to this:

```csharp
public EditContext? EditContext
{
    get => _fixedEditContext;
    set
    {
        _fixedEditContext = value;
        _hasSetEditContextExplicity = value != null ? true : false;
    }
}
```

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/EditForm.cs:46]: Looks like that would work to me.

Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs]: Ok I've push the update

File: src/Components/Web/src/Forms/EditForm.cs
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/EditForm.cs:91]: Should this now be `else if (!_hasSetEditContextExplicitly && Model == null)` ?

Because the rule is: you must supply a model if and only if you didn't supply an edit context.

In the event where you:

 * First supplied a non-null model
 * Then on the next render supplied a null model

Then we'd want to throw I think.

File: src/Components/Web/src/Forms/EditForm.cs
Comment (SteveSandersonMS) [inline src/Components/Web/src/Forms/EditForm.cs:109]: Nit: I think this could be simplified to:

```suggestion
            if (!_hasSetEditContextExplicitly && Model != _fixedEditContext?.Model)
```

...  because if `_hasSetEditContextExplicitly` then we know `_fixedEditContext != null`.

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/EditForm.cs:109]: Also the comment above it is slightly out of date now

Comment (SteveSandersonMS) [reply src/Components/Web/src/Forms/EditForm.cs:109]: Strictly speaking the check could be expressed as:

```cs
if (Model != null && Model != _fixedEditContext?.Model)
```

(because we know that if `Model != null`, then `_hasSetEditContextExplicitly` is false otherwise we would already have thrown above)

Arguably this is a more natural phrasing because it's just stating that if you give us any new model, we want to use it.

**Comment by SteveSandersonMS** [comment]
Thanks very much for the updates, @chrissainty! I think this is really close to being done now.

One final thought: the name `_fixedEditContext` no longer really makes sense, and we could simplify it just to being called `_editContext` now. It's simply the backing field for the `EditContext` property now. If you feel like making this rename then great, but if not that's absolutely fine and we won't hold up the PR on it :)

**Comment by chrissainty** [comment]
Give me 2 minutes @SteveSandersonMS, I'll get it done! :)

File: src/Components/Web/src/Forms/EditForm.cs
Comment (chrissainty) [reply src/Components/Web/src/Forms/EditForm.cs:91]: Yes, agreed. I'll update it.

**Comment by SteveSandersonMS** [comment]
Awesome, thanks!

@captainsafia @pranavkm Can I leave it with one of you to merge when the tests pass and if you're happy with it? Up to you whether you also want to ask for/wait for [this update too](https://github.com/dotnet/aspnetcore/pull/24007#discussion_r456008624).

**Comment by chrissainty** [comment]
@SteveSandersonMS @captainsafia @pranavkm I've addressed the last comment, all should be good now :)

**Comment by captainsafia** [comment]
> @SteveSandersonMS @captainsafia @pranavkm I've addressed the last comment, all should be good now :)

Thanks, Chris! Taking a look now.

File: src/Components/Web/test/Forms/EditFormTest.cs
Comment (captainsafia) [reply src/Components/Web/test/Forms/EditFormTest.cs:64]: Resolving this as it looks like @chrissainty has already made this update!

**Review by captainsafia — APPROVED** [review]
Looks like all the feedback has ben addressed here. Thanks for adding the tests!

**Review by SteveSandersonMS — APPROVED** [review]
Brilliant. Thanks @chrissainty!

**Comment by chrissainty** [comment]
My pleasure! This first sprint has been awesome, I hope we can do it again, soon 😀

**Comment by captainsafia** [comment]
@chrissainty Thank you for the contribution! Your pull request has been merged!

**Pull request merged**
